File: captainhcg/py/315-count-of-smaller-numbers-after-self.py
class Node(object):
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.count = 0
        self.smaller = 0

class Solution(object):
    def countSmaller(self, nums):
        if not nums:
            return []

        root = Node(float("inf"))
        ret = []

        def insert(node, v, cnt):
            if v == node.val:
                ret.append(cnt + node.smaller)
                node.count += 1
            elif v < node.val:
                node.smaller += 1
                if not node.left:
                    node.left = Node(v)
                insert(node.left, v, cnt)
            elif v > node.val:
                cnt += node.smaller + node.count
                if not node.right:
                    node.right = Node(v)
                insert(node.right, v, cnt)

        for n in nums[::-1]:
            insert(root, n, 0)

        return ret[::-1]

File: captainhcg/js/315-count-of-smaller-numbers-after-self.js
var countSmaller = function(nums) {
    var arr = []
    nums.forEach(function(n, i){
        arr.push({val: n, idx: i, smaller: 0})
    });
    var tmpArr = [];
    var mergeSort = function(start, end){
        if(start >= end){
            return
        }
        var mid = Math.floor((start + end) / 2);
        mergeSort(start, mid);
        mergeSort(mid+1, end);
        var left = start, right = mid + 1, ptr = 0;
        while (left <= mid && right <= end){
            if (arr[left].val > arr[right].val){
                arr[left].smaller += end - right + 1
                tmpArr[ptr++] = arr[left++];
            } else {
                tmpArr[ptr++] = arr[right++];
            }
        }
        while (left <= mid){
            tmpArr[ptr++] = arr[left++];
        }
        while (right <= end){
            tmpArr[ptr++] = arr[right++];
        }
        for(var idx = 0; idx < end - start + 1; idx++){
            arr[start + idx] = tmpArr[idx];
        }
    }
    mergeSort(0, nums.length-1)
    ret = Array(nums.length);
    arr.forEach(function(obj){
        ret[obj.idx] = obj.smaller;
    });
    return ret;
};
